.github/scripts/workflow-monitor.py

```diff
@@ -32,44 +35,71 @@
 
-def main(platform: Platform, workflow_id: str, gha_ci_personal_token: str):
+def wrap_in_code(wrap: str)
+    return f"\n```\n{wrap}\n```"
 
+def main(platform: Platform):
     consecutive_failures = 0
-    headers = {'Authorization': "token {}".format(gha_ci_personal_token.strip())}
-    gha_workflow_api_url = f"{gha_runs_api_url}/{workflow_id}"
 
+    print("Workflow monitor started")
     platform_lib = get_platform_lib(platform)
-    while True:
-        time.sleep(POLLING_INTERVAL_SECONDS)
-
-        res = requests.get(gha_workflow_api_url, headers=headers)
-        if res.status_code == 200:
-            consecutive_failures = 0
-            res_dict = res.json()
-            state_status = res_dict["status"]
-            state_concl = res_dict["conclusion"]
-
-            print("Workflow {} status: {} {}".format(workflow_id, state_status, state_concl))
-            if state_status in ['completed']:
-                if state_concl in TERMINATE_STATES:
-                    platform_lib.terminate_instances(gha_ci_personal_token, workflow_id)
-                    exit(0)
-                elif state_concl in STOP_STATES:
-                    platform_lib.stop_instances(gha_ci_personal_token, workflow_id)
-                    exit(0)
-                elif state_concl not in NOP_STATES:
-                    print("Unexpected Workflow State On Completed: {}".format(state_concl))
+
+    try:
+        print("Sending startup message")
+        issue_post(ci_env['PERSONAL_ACCESS_TOKEN'], f"Workflow monitor started for CI run: {ci_env['GITHUB_RUN_ID']}")
+
+        print("Beginning polling loop")
+        while True:
+            time.sleep(POLLING_INTERVAL_SECONDS)
+
+            res = requests.get(gha_workflow_api_url, headers=get_header(ci_env['PERSONAL_ACCESS_TOKEN']))
+            if res.status_code == 200:
+                consecutive_failures = 0
+                res_dict = res.json()
+                state_status = res_dict["status"]
+                state_concl = res_dict["conclusion"]
+
+                print(f"Workflow {ci_env['GITHUB_RUN_ID']} status: {state_status} {state_concl}")
+
+                # check that select instances are terminated on time
+                platform_lib.check_and_terminate_select_instances(45, ci_env['GITHUB_RUN_ID'])
+
+                if state_status in ['completed']:
+                    if state_concl in TERMINATE_STATES:
+                        platform_lib.terminate_instances(ci_env['PERSONAL_ACCESS_TOKEN'], ci_env['GITHUB_RUN_ID'])
+                        return
+                    elif state_concl in STOP_STATES:
+                        platform_lib.stop_instances(ci_env['PERSONAL_ACCESS_TOKEN'], ci_env['GITHUB_RUN_ID'])
+                        return
+                    elif state_concl not in NOP_STATES:
+                        print(f"Unexpected Workflow State On Completed: {state_concl}")
+                        raise ValueError
```

Comment:
Since you're catching exceptions the exceptions here, it might be better to raise an expression with the print message inside of it so that it can be seen from the PR comment, otherwise you're going to have a value error raised and follow the stack trace to see what happened. Something like `raise ValueError(f"Unexpected Workflow State on Completed: {state_concl}")` etc.

Reply:
Done